packet.go

```diff
@@ -66,29 +89,70 @@ func (p *Packet) IsRaw() bool {
 
 // Decode decodes the headers of a Packet.
 func (p *Packet) Decode() {
+
+	// Our go tests do not initialize p.Payload.
+	if p.Payload == nil {
+		p.Payload = p.Data
+	}
 	if p.IsRaw() == false {
-		p.Type = int(binary.BigEndian.Uint16(p.Data[12:14]))
-		p.DestMac = decodemac(p.Data[0:6])
-		p.SrcMac = decodemac(p.Data[6:12])
-		p.Payload = p.Data[14:]
-		switch p.Type {
-		case 0x8100:
-			//VLAN tag skip 802.1Q tag
-			p.Type = int(binary.BigEndian.Uint16(p.Data[16:18]))
-			p.Payload = p.Data[18:]
-		case 0x88A8:
-			//VLAN tag skip 802.1ad double tag
-			p.Type = int(binary.BigEndian.Uint16(p.Data[20:22]))
-			p.Payload = p.Data[22:]
-		}
+		// we want to set mac addresses of the Packet p using outermost
+		// ethernet frame. These are used to determine directionality of the
+		// captured packet.
+		p.decodeEther(true /* outerMac */)
 	} else {
 		p.Type = TYPE_IP
-		if p.Data[0]&0xf0 == 6 {
+		if p.Payload[0]&0xf0 == 6 {
 			p.Type = TYPE_IP6
 		}
-		p.Payload = p.Data[0:]
+		p.Payload = p.Payload[0:]
+		p.decodeL3()
+	}
+}
+
+// returns false if no valid ethernet header is found; true otherwise.
+// Currently, we need return value for vxland decoding.
+func (p *Packet) decodeEther(outerMac bool) bool {
+	tData := p.Payload
+	tData_len := len(tData)
+	if tData_len < 14 {
+		return false
+	}
+	p.Type = int(binary.BigEndian.Uint16(tData[12:14]))
+	if outerMac {
+		p.DestMac = decodemac(tData[0:6])
+		p.SrcMac = decodemac(tData[6:12])
+	}
+	p.Payload = tData[14:]
+	switch p.Type {
+	case 0x8100:
+		//VLAN tag skip 802.1Q tag
+		if tData_len < 18 {
+			return false
+		}
+		p.Type = int(binary.BigEndian.Uint16(tData[16:18]))
+		p.Payload = tData[18:]
+	case 0x88A8:
+		//VLAN tag skip 802.1ad double tag
+		if tData_len < 22 {
+			return false
+		}
+		p.Type = int(binary.BigEndian.Uint16(tData[20:22]))
+		p.Payload = tData[22:]
 	}
 
+	p.decodeL3()
+	return true
+}
+
+// this function assumes p.Type is set to one of three supported L2.5/L3
+// protocol types i.e. IP, IP6 or ARP. For other protocols, we don't do
```

Comment:
It looks like we do not decodeArp as well although we identify the type in line 97

Reply:
Looks like there is some past reason to ignore ARP. We drop arp in agent-repo too.